l3kernel/l3clist.dtx

```diff
@@ -351,6 +351,23 @@
 %   described in Section~\ref{sec:l3sort:mech}.
 % \end{function}
 %
+% \begin{function}[added = 2021-01-23, EXP]{\clist_sort:NN, \clist_sort:nN}
+%   \begin{syntax}
+%     \cs{clist_sort:NN} \meta{clist var} \meta{conditional}
+%   \end{syntax}
+%   Sorts the items in the \meta{clist var}, using the \meta{conditional} to
+%   compare the items, and leaves the result in the input stream as a |clist|.
+%   The \meta{conditional} should have signature |:nnTF|, and return
+%   \texttt{true} if the two items being compared should be left in the same
+%   order, and \texttt{false} if the items should be swapped. The details of
+%   sorting comparison are described in Section~\ref{sec:l3sort:mech}.
+%   \begin{texnote}
+%     The result is returned within \cs{exp_not:n}, which means that the
+%     token list does not expand further when appearing in an
+%     \texttt{x}-type or \texttt{e}-type argument expansion.
+%   \end{texnote}
+% \end{function}
+%
```

Comment:
We have a non-expandable `\tl_reverse:N` modifying a tl variable, and an expandable `\tl_reverse:n` that leaves its result in the input stream.  To do the second (expandable) operation on the contents of a token list variable, we use `\tl_reverse:V`, which is not a synonym of `\tl_reverse:N`.

For the same reason we shouldn't implement an expandable `\clist_sort:NN` (clist var) (comparison) that would be likely to get confused with the non-expandable assignment `\clist_sort:Nn (clist var) (comparison)`.  Instead we should implement the variant `\clist_sort:VN` of `\clist_sort:nN`.

Alternatively, it might be better to have `\clist_sorted` for the expandable version, like Python's `sorted(l)` versus `l.sort()`, but perhaps the ship has sailed already given the expandable `\tl_sort:n` and the expandable `\tl_reverse:n` (which with this logic should be `\tl_reversed:n`).

Reply:
The reasoning I had to not call it `\clist_sort:VN` was that it doesn't do a `V`-type expansion, but only works correctly for a `clist`-variable (with `V` one could expect it to work with a `tl` as well, but this would only really do the expected thing if that `tl` was also a correctly sanatised `clist` just somehow stored in a `tl`).

But I'm happy with whatever name you pick.